### PR TITLE
Add ERC1820 Registry as a precompile

### PR DESCRIPTION
**Pull request by dmihal**
## Description
Multiple standards, most notably ERC-777, depend on ERC-1820 to act as a registry of contract interfaces.

The ERC-1820 standard depends on the registry contract being deployed at a defined address (`0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24`). This is achieved using "Nick's method", in which the resulting address is reverse-engineered from a generated signature.

Unfortunately, this technique is not possible on Optimism, since the OVM bytecode is different than the EVM bytecode.

Therefore, the best option to support ERC-1820 is to add this registry contract as a precompile deployed to the canonical address.

## Contributing Agreement

- [X] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

**Comment by tynes** [comment]
This will fix the broken integration tests: https://github.com/ethereum-optimism/contracts-v2/commit/160bd939435d8b2d76a9bea9ca56eb3aa2bd24cc

**Comment by maurelian** [comment]
Thank you for this @dmihal, and my apologies for the delayed response.

The dependency on Nick's method is a very interesting challenge. Otherwise it would be easy to say that this should simply be deployed once we have our open mainnet release live. 

How important is it to the standard that this address be maintained? 
Are you aware of other such standards that we should consider for 'predeployment'?

**Comment by maurelian** [comment]
FYI @dmihal, we're going to move forward with including this. LGTM on first glance, but definitely would like to get a review on it from @ben-chain.

File: contracts/optimistic-ethereum/OVM/precompiles/ERC1820Registry.sol [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+ *    ███████╗██║  ██║╚██████╗ ██║╚█████╔╝███████╗╚██████╔╝
+ *    ╚══════╝╚═╝  ╚═╝ ╚═════╝ ╚═╝ ╚════╝ ╚══════╝ ╚═════╝
+ *
+ *    ██████╗ ███████╗ ██████╗ ██╗███████╗████████╗██████╗ ██╗   ██╗
+ *    ██╔══██╗██╔════╝██╔════╝ ██║██╔════╝╚══██╔══╝██╔══██╗╚██╗ ██╔╝
+ *    ██████╔╝█████╗  ██║  ███╗██║███████╗   ██║   ██████╔╝ ╚████╔╝
+ *    ██╔══██╗██╔══╝  ██║   ██║██║╚════██║   ██║   ██╔══██╗  ╚██╔╝
+ *    ██║  ██║███████╗╚██████╔╝██║███████║   ██║   ██║  ██║   ██║
+ *    ╚═╝  ╚═╝╚══════╝ ╚═════╝ ╚═╝╚══════╝   ╚═╝   ╚═╝  ╚═╝   ╚═╝
+ *
```
Comment (maurelian) [inline contracts/optimistic-ethereum/OVM/precompiles/ERC1820Registry.sol:27]: I'd prefer to remove this ascii-art. :)

**Comment by tynes** [comment]
The failing test is due to the PR coming from a fork and the docker stuff being built to use position independent dockerfiles, meaning the dockerfile pulls the repo as part of the docker build. This should fix it here: https://github.com/ethereum-optimism/docker/pull/28 - just need to take the time to test this branch to make sure its backwards compatible

**Comment by dmihal** [comment]
I've updated & rebased this PR, but still seeing that test failing. @tynes let me know if there's anything I should do to to incorporate that fix.

**Comment by tynes** [comment]
Hey @dmihal, this change should fix the problem: https://github.com/ethereum-optimism/contracts-v2/commit/791370b8b1ced9ab27c472983b77eaae2517818a
On a rebase it should build for the tests properly. Apologies for it not just working.

Convo in another repo: https://github.com/ethereum-optimism/data-transport-layer/pull/49

**Comment by dmihal** [comment]
@tynes hmm seems to still be failing after the rebase

**Review by tynes — APPROVED** [review]

**Comment by ben-chain** [comment]
Thanks again for this @dmihal; sorry for the slow merge, amazing stuff!!! :)

**Pull request merged**